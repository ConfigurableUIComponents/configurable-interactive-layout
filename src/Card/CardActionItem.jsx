import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { composeThemeFromProps } from '@css-modules-theme/react';

import buildThemeProperties from '../Utils';

import style from '../Layout/LayoutStyle.scss';

export default class CardActionItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      hover: false,
    };
    this.action = props.action;
    this.onClickAction = this.assignOnClickAction();
  }

  onMouseItem(hover) {
    this.setState({ hover });
  }

  getActionIconURL() {
    const { hover } = this.state;
    const { defaultIcon } = this.props;

    if (!hover && this.action.iconURL) {
      return this.action.iconURL;
    }
    if (hover && this.action.iconURLHover) {
      return this.action.iconURLHover;
    }
    if (hover && this.action.iconURLHover === undefined && this.action.iconURL) {
      return this.action.iconURL;
    }
    return defaultIcon;
  }

  assignOnClickAction() {
    let onClickAction = this.action.onClick;
    if (onClickAction === undefined) {
      onClickAction = this.publishActionEvent;
    }
    return onClickAction;
  }

  publishActionEvent(actionId) {
    const { cardId, eventManager } = this.props;
    eventManager.publish(actionId, cardId, {});
  }

  render() {
    const { theme, themeProps } = this.props;
    const themingProperties = buildThemeProperties(theme, themeProps);
    const themeStyles = composeThemeFromProps(style, themingProperties, {
      compose: 'Merge',
    });
    const { onClickAction } = this;
    const actionIcon = this.getActionIconURL();
    // const divStyle = { background: `url('${actionIcon}') no-repeat center center` };
    const actionTooltip = (this.action.displayName) ? ` ${themeStyles.actionTooltip}` : null;
    return (
      <div
        className={`${themeStyles.cardActionItem}${actionTooltip}`}
        data-tooltip={this.action.displayName}
        key={this.action.id}
        onClick={onClickAction}
        onKeyUp={onClickAction}
        tabIndex="0"
        role="button"
        onMouseOver={() => this.onMouseItem(true)}
        onFocus={() => this.onMouseItem(true)}
        onMouseOut={() => this.onMouseItem(false)}
        onBlur={() => this.onMouseItem(false)}
      >
        <img src={actionIcon} alt={this.action.displayName} />
      </div>
    );
  }
}


CardActionItem.propTypes = {
  cardId: PropTypes.string.isRequired,
  action: PropTypes.shape({
    id: PropTypes.string,
    displayName: PropTypes.string,
    iconURL: PropTypes.string,
    iconURLHover: PropTypes.string,
  }),
  defaultIcon: PropTypes.string,
  eventManager: PropTypes.instanceOf(Object),
  themeProps: PropTypes.shape({
    compose: PropTypes.string,
    prefix: PropTypes.string,
  }),
  theme: PropTypes.instanceOf(Object),
};

CardActionItem.defaultProps = {
  themeProps: {
    compose: 'merge',
    prefix: 'configurableInteractiveLayout-',
  },
};
