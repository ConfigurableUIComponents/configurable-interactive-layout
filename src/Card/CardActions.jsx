import React, { Component } from 'react';
import PropTypes from 'prop-types';
import each from 'lodash/each';
import cloneDeep from 'lodash/cloneDeep';

import CardActionItem from './CardActionItem';


const DEFAULT_ACTION_ICON = 'SV_ANN.svg';

export default class CardActions extends Component {
  constructor(props) {
    super(props);
    this.state = {
      menuOpen: false,
    };
    if (this.props.actions) {
      this.items = this.buildActionItems();
    }
  }

  onActionMenuClick() {
    this.setState({ menuOpen: !this.state.menuOpen });
  }

  buildActionItems() {
    const actionControls = [];
    each(this.props.actions, (originalAction, actionId) => {
      const action = cloneDeep(originalAction);
      action.id = actionId;
      if (!action.iconURL && !action.displayName) {
        return null;
      }
      actionControls.push(<CardActionItem
        key={actionId}
        action={action}
        defaultIcon={DEFAULT_ACTION_ICON}
        cardId={this.props.cardId}
        eventManager={this.props.eventManager}
      />);
    });
    return actionControls;
  }

  render() {
    const actionItemsClassName = `card-action-items ${this.state.menuOpen ? '' : 'dispnone'}`;
    const actionsMenuClassName = `cards-actions-menu ${this.state.menuOpen ? 'expanded' : 'collapsed'}`;
    const actionsClassName = `actions ${this.state.menuOpen ? 'menu-open' : ''}`;

    if (this.items.length > 0) {
      return (
        <div className={actionsClassName} >
          <div
            className={actionsMenuClassName}
            onClick={() => this.onActionMenuClick()}
            onKeyUp={() => this.onActionMenuClick()}
            role="button"
            tabIndex={0}
          />
          <div className={actionItemsClassName} > { this.items } </div>
        </div>);
    }
    return null;
  }
}


CardActions.propTypes = {
  cardId: PropTypes.string.isRequired,
  actions: PropTypes.instanceOf(Object),
  eventManager: PropTypes.instanceOf(Object),
};
