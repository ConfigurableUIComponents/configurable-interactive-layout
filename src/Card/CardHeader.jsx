import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CardActions from './CardActions';

export default class CardHeader extends Component {
  createHeaderClassName() {
    return `card-header ${this.props.title ? '' : 'without-title'}`;
  }

  render() {
    return (
      <div className={this.createHeaderClassName()}>
        {this.props.actions ? <CardActions {...this.props} /> : null }
        {this.props.title ? <div className="title" title={this.props.title}>{this.props.title}</div> : null }
      </div>);
  }
}

CardHeader.propTypes = {
  cardId: PropTypes.string.isRequired,
  title: PropTypes.string,
  actions: PropTypes.instanceOf(Object),
  eventManager: PropTypes.instanceOf(Object),
};

CardHeader.defaultProps = {
  title: undefined,
  actions: undefined,
  eventManager: undefined,
};
