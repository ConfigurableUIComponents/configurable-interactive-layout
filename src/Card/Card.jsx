import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CardHeader from './CardHeader';

const NEW_CARD_HIGHLIGHT_TIMEOUT = 1000;

export default class Card extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isNewCard: props.isNewCard,
    };
  }

  componentDidMount = () => {
    if (this.newCardRef) {
      this.newCardRef.current.scrollIntoView();
      this.newCardRef = undefined;
      setTimeout(() => {
        this.setState({ isNewCard: false });
      }, NEW_CARD_HIGHLIGHT_TIMEOUT);
    }
  }

  componentWillReceiveProps = (nextProps) => {
    this.setState({ isNewCard: nextProps.isNewCard });
  }

  render() {
    let showHeader = true;
    let cardClassName = 'card ';
    if (this.props.cardCssClass) {
      cardClassName += this.props.cardCssClass;
    }
    let cardRef;
    if (!this.props.title && !this.props.actions) showHeader = false;
    else if (this.props.title) {
      cardClassName += ' with-title';
    }
    if (this.state.isNewCard && React.createRef) {
      cardRef = React.createRef();
      this.newCardRef = cardRef; // we will use this reference to scroll the view to this card
      cardClassName += ' highlight';
    }
    return (
      <div ref={cardRef} className={cardClassName}>
        {showHeader ?
          <CardHeader
            cardId={this.props.configId}
            {...this.props}
          />
          : null}
        <div className="card-content">
          { this.props.children }
        </div>
      </div>
    );
  }
}

Card.propTypes = {
  configId: PropTypes.string.isRequired,
  title: PropTypes.string,
  cardCssClass: PropTypes.string,
  isNewCard: PropTypes.bool,
  store: PropTypes.instanceOf(Object),
  listeners: PropTypes.instanceOf(Array),
  eventManager: PropTypes.instanceOf(Object),
  actions: PropTypes.instanceOf(Object),
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]).isRequired,
};

Card.defaultProps = {
  title: undefined,
  actions: undefined,
  eventManager: undefined,
  store: undefined,
};
