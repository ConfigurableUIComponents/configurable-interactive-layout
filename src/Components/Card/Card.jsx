import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CardHeader from './CardHeader';

export default class Card extends Component {
  componentWillUnmount() {
    // this.props.listeners.map(event => this.props.eventManager.unsubscribe(event.id));
  }

  render() {
    let showHeader = true;
    let cardClassName = 'card';
    if (!this.props.title && !this.props.actions) showHeader = false;
    else if (this.props.title) {
      cardClassName += ' with-title';
    }
    return (
      <div className={cardClassName}>
        {showHeader ?
          <CardHeader
            cardId={this.props.configId}
            {...this.props}
          />
          : null}
        <div className="card-content">
          { this.props.children }
        </div>
      </div>
    );
  }
}

Card.propTypes = {
  configId: PropTypes.string.isRequired,
  title: PropTypes.string,
  store: PropTypes.instanceOf(Object),
  listeners: PropTypes.instanceOf(Array),
  eventManager: PropTypes.instanceOf(Object),
  actions: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    displayName: PropTypes.string,
  })),
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]).isRequired,
};

Card.defaultProps = {
  title: undefined,
  actions: undefined,
  eventManager: undefined,
  store: undefined,
};
