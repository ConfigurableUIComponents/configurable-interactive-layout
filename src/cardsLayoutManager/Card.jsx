import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CardHeader from './CardHeader';

export default class Card extends Component {
  render() {
    return (
      <div className="card">
        {
          <CardHeader
            {...this.props}
          /> }
        <div className="card-content">
          { this.props.children }
        </div>
      </div>
    );
  }
}

Card.propTypes = {
  title: PropTypes.string,
  store: PropTypes.instanceOf(Object),
  eventManager: PropTypes.instanceOf(Object),
  actions: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    displayName: PropTypes.string,
  })),
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]).isRequired,
};

Card.defaultProps = {
  title: undefined,
  actions: undefined,
  eventManager: undefined,
  store: undefined,
};
