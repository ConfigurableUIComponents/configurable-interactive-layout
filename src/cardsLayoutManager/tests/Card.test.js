import React from 'react';
import Card from '../Card';
// import renderer from 'react-test-renderer';

describe('Card testing', () => {
  describe('Card initialization', () => {
    it('Children props are not empty', () => {
      });
    });
  });
});

