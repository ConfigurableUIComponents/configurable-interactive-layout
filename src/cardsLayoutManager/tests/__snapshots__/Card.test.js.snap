// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card changes the class when hovered 1`] = `
<div
  className="card"
>
  <div
    className="card-content"
  >
    Facebook
  </div>
</div>
`;
